File: x/psets/1/credit/credit.c
/*
    JONASH MARCELINO
    CS50
*/
#include <cs50.h>
#include <stdio.h>

// Functions Declaration
bool check_card(long);
int get_card_length(long);
int get_first_two_digits(long);

int main(void)
{
    // TODO: Promt for input
    long card_no;
    do
    {
        card_no = get_long("Number: ");
    }
    while (card_no < 1);

    // TODO: Calculate checksum
    bool isValid = check_card(card_no);

    // TODO: Check card length and starting digits.
    int ccLength = get_card_length(card_no);
    int digits = get_first_two_digits(card_no);

    // TODO: Print AMEX, MASTERCARD, VISA, or INVALID.
    if (!isValid)
    {
        printf("INVALID\n");
    }
    else if (digits / 10 == 4 && (ccLength == 16 || ccLength == 13))
    {
        printf("VISA\n");
    }
    else if ((digits == 34 || digits == 37) && ccLength == 15)
    {
        printf("AMEX\n");
    }
    else if ((digits >= 51 && digits <= 55) && ccLength == 16)
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

// Function to calculate checksum.
bool check_card(long num)
{
    int sum = 0;
    while (num > 0)
    {
        int lastDigit;
        // Get the last digit
        lastDigit = num % 10;
        // Add the last digit to sum
        sum += lastDigit;
        // Move to next digit
        num /= 10;
        // Get the last digit
        lastDigit = num % 10;
        // Multiply the last digit
        int timesTwo = lastDigit * 2;
        if (timesTwo > 9)
        {
            // Add the two digits to the sum not the number
            sum += (timesTwo % 10) + (timesTwo / 10);
        }
        else
        {
            // Add the multiplied number
            sum += timesTwo;
        }
        // Move to next digit
        num /= 10;
    }
    if (sum % 10 == 0)
    {
        return true;
    }
    return false;
}

// Function to check card length.
int get_card_length(long num)
{
    int length = 0;
    while (num > 0)
    {
        length++;
        num /= 10;
    }
    return length;
}

// Function to check card starting digits.
int get_first_two_digits(long num)
{
    int length = get_card_length(num);
    // We divide the num to 10 to finally to 2 digits only
    for (int i = length; i > 2; i--)
    {
        num /= 10;
    }
    return num;
}

File: x/psets/6/sentimental-credit/credit.py
# TODO
from cs50 import get_int


def main():
    # Get input from user
    while True:
        card_no = get_int("Number: ")
        if card_no > 1:
            break

    # Check the card_no if valid
    is_valid = check_card(card_no)
    # Int to string
    cc_str = str(card_no)
    # Get length
    length = len(cc_str)
    # Get two digits
    first_two = cc_str[0] + cc_str[1]

    # Conditions for printing
    if not is_valid:
        print("INVALID")
    elif first_two[0] == '4' and length == 16 or length == 13:
        print("VISA")
    elif first_two == '34' or first_two == '37' and length == 15:
        print("AMEX")
    elif int(first_two) >= 34 and int(first_two) <= 55 and length == 16:
        print("MASTERCARD")
    else:
        print("INVALID")


def check_card(ccn):
    """ Check the card's validity """
    sum = 0
    while ccn > 0:
        last_digit = ccn % 10
        sum += last_digit
        ccn = int(ccn / 10)
        last_digit = ccn % 10
        times_two = last_digit * 2
        if times_two > 9:
            sum += (times_two % 10) + (int(times_two / 10))
        else:
            sum += times_two
        ccn = int(ccn / 10)
    if sum % 10 == 0:
        return True
    return False


if __name__ == '__main__':
    main()